File: app/src/main/java/com/kotobyte/searchpage/SearchPageFragment.java
package com.kotobyte.searchpage;

import android.databinding.DataBindingUtil;
import android.os.Bundle;
import android.support.annotation.Nullable;
import android.support.v4.app.Fragment;
import android.util.Log;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.Toast;

import com.kotobyte.R;
import com.kotobyte.base.Configuration;
import com.kotobyte.base.ServiceLocator;
import com.kotobyte.databinding.FragmentSearchPageBinding;
import com.kotobyte.models.Kanji;
import com.kotobyte.models.Word;

import java.util.List;

public class SearchPageFragment extends Fragment implements SearchPageContracts.View {

    public static final String TAG = SearchPageFragment.class.getSimpleName();

    private static final String ARG_QUERY = "query";

    private FragmentSearchPageBinding mBinding;
    private SearchPageDataSource mDataSource;
    private SearchPagePresenter mPresenter;
    private SearchResultsAdapter mSearchResultsAdapter;

    public static SearchPageFragment newInstance(CharSequence query) {

        SearchPageFragment fragment = new SearchPageFragment();

        Bundle arguments = new Bundle();
        arguments.putString(ARG_QUERY, query.toString());

        fragment.setArguments(arguments);
        return fragment;
    }

    public String getQuery() {
        return getArguments().getString(ARG_QUERY);
    }

    @Override
    public void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);

        Configuration configuration = ServiceLocator.getInstance().getConfiguration();

        mDataSource = new SearchPageDataSource(configuration.getDictionaryFilePath());
        mPresenter = new SearchPagePresenter(this, mDataSource, getQuery());
    }

    @Nullable
    @Override
    public View onCreateView(LayoutInflater inflater, @Nullable ViewGroup container, Bundle savedInstanceState) {
        mBinding = DataBindingUtil.inflate(inflater, R.layout.fragment_search_page, container, false);

        return mBinding.getRoot();
    }

    @Override
    public void onActivityCreated(@Nullable Bundle savedInstanceState) {
        super.onActivityCreated(savedInstanceState);

        mPresenter.onCreate();
    }

    @Override
    public void onDestroy() {
        super.onDestroy();

        mPresenter.onDestroy();
        mDataSource.close();
    }

    @Override
    public void showWordSearchProgressBar(boolean show) {
        mBinding.progressBar.setVisibility(show ? View.VISIBLE : View.GONE);
    }

    @Override
    public void showWordSearchResultsView(boolean show) {
        mBinding.searchResultsView.animate().alpha(show ? 1 : 0).start();
    }

    @Override
    public void showWordSearchResults(List<Word> words) {
        mSearchResultsAdapter = new SearchResultsAdapter(getContext(), mWordSearchResultsAdapterListener, words);

        mBinding.searchResultsView.setAdapter(mSearchResultsAdapter);
    }

    @Override
    public void showNoWordSearchResultsLabel(boolean show) {
        mBinding.noSearchResultsLabel.setVisibility(show ? View.VISIBLE : View.GONE);
    }

    @Override
    public void showKanjiSearchResults(int position, List<Kanji> kanjiList) {
        mSearchResultsAdapter.setKanjiSearchResults(position, kanjiList);
    }

    @Override
    public void showKanjiDetailScreen(Kanji kanji) {
        KanjiDetailDialogFragment.newInstance(kanji).show(getChildFragmentManager(), KanjiDetailDialogFragment.TAG);
    }

    @Override
    public void showUnknownError(Throwable error) {
        Log.e(TAG, error.getLocalizedMessage(), error);

        Toast.makeText(getContext(), R.string.common_unknown_error, Toast.LENGTH_SHORT).show();
    }

    private SearchResultsAdapter.Listener mWordSearchResultsAdapterListener = new SearchResultsAdapter.Listener() {

        @Override
        public void onRequestKanjiListForWord(int position, Word word) {
            mPresenter.onRequestKanjiListForWord(position, word);
        }

        @Override
        public void onRequestDetailForKanji(Kanji kanji) {
            mPresenter.onRequestDetailForKanji(kanji);
        }
    };
}

File: app/src/main/java/com/kotobyte/search/SearchPageFragment.kt
package com.kotobyte.search

import android.databinding.DataBindingUtil
import android.os.Bundle
import android.support.v4.app.Fragment
import android.util.Log
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup

import com.kotobyte.R
import com.kotobyte.base.ServiceLocator
import com.kotobyte.databinding.FragmentSearchPageBinding
import com.kotobyte.models.Kanji
import com.kotobyte.models.Word
import com.kotobyte.utils.ErrorDialogFragment

class SearchPageFragment : Fragment(), SearchPageContracts.View {

    private var binding: FragmentSearchPageBinding? = null
    private var presenter: SearchPageContracts.Presenter? = null
    private var searchResultsAdapter: SearchResultsAdapter? = null

    val query: String
        get() = arguments?.getString(ARG_QUERY) ?: ""

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)

        presenter = SearchPagePresenter(this, SearchPageDataSource(ServiceLocator.databaseProvider!!), query)
    }

    override fun onCreateView(inflater: LayoutInflater?, container: ViewGroup?, savedInstanceState: Bundle?): View? {
        binding = DataBindingUtil.inflate(inflater!!, R.layout.fragment_search_page, container, false)

        return binding?.root
    }

    override fun onActivityCreated(savedInstanceState: Bundle?) {
        super.onActivityCreated(savedInstanceState)

        presenter?.onCreate()
    }

    override fun onDestroy() {
        super.onDestroy()

        presenter?.onDestroy()
    }

    override fun showWordSearchProgressBar(show: Boolean) {
        binding?.progressBar?.visibility = if (show) View.VISIBLE else View.GONE
    }

    override fun showWordSearchResultsView(show: Boolean) {
        binding?.searchResultsView?.animate()?.alpha(if (show) 1f else 0f)?.start()
    }

    override fun showWordSearchResults(words: List<Word>) {
        searchResultsAdapter = SearchResultsAdapter(context, mWordSearchResultsAdapterListener, words)

        binding?.searchResultsView?.adapter = searchResultsAdapter
    }

    override fun showNoWordSearchResultsLabel(show: Boolean) {
        binding?.noSearchResultsLabel?.visibility = if (show) View.VISIBLE else View.GONE
    }

    override fun showKanjiSearchResults(position: Int, kanjiList: List<Kanji>) {
        searchResultsAdapter?.setKanjiSearchResults(position, kanjiList)
    }

    override fun showKanjiDetailScreen(kanji: Kanji) =
            KanjiDetailDialogFragment.create(kanji).show(childFragmentManager, KanjiDetailDialogFragment::class.java.simpleName)

    override fun showUnknownError(error: Throwable) {
        Log.e(TAG, error.localizedMessage, error)

        ErrorDialogFragment.create().show(fragmentManager, ErrorDialogFragment::class.java.simpleName)
    }

    private val mWordSearchResultsAdapterListener = object : SearchResultsAdapter.Listener {

        override fun onRequestKanjiListForWord(position: Int, word: Word) {
            presenter?.onRequestKanjiListForWord(position, word)
        }

        override fun onRequestDetailForKanji(kanji: Kanji) {
            presenter?.onRequestDetailForKanji(kanji)
        }
    }

    companion object {
        private val TAG = SearchPageFragment::class.java.simpleName

        private val ARG_QUERY = "query"

        fun create(query: CharSequence) = SearchPageFragment().apply {

            arguments = Bundle().apply {
                putString(ARG_QUERY, query.toString())
            }
        }
    }
}
